src/utils.rs

```diff
@@ -152,6 +153,77 @@ pub fn ensure_procfs(path: &Path) -> Result<()> {
     Ok(())
 }
 
+pub fn secure_join(rootfs: &Path, unsafe_path: &Path) -> Result<PathBuf> {
+    let mut clean_path = PathBuf::from(rootfs);
+    let mut path = PathBuf::from(unsafe_path);
+    let mut p = PathBuf::new();
+
+    let mut part = path.iter();
+    let mut i = 0;
+
+    loop {
+        if i > 255 {
+            // Dereference too many symlinks, may be infinite loop
+            bail!("dereference too many symlinks, may be infinite loop");
+        }
+
+        let part_path;
+        match part.next() {
+            Some(part) => {
+                part_path = PathBuf::from(part);
+            }
+            None => {
+                break;
+            }
+        }
+
+        if !part_path.is_absolute() {
+            if part_path.starts_with("..") {
+                p.pop();
+            } else {
+                // check if symlink then dereference
+                let metadata = match PathBuf::from(&clean_path)
+                    .join(&p)
+                    .join(&part_path)
+                    .symlink_metadata()
+                {
+                    Ok(metadata) => Some(metadata),
+                    Err(error) => match error.kind() {
+                        // if file does not exists, treat it as normal path
+                        ErrorKind::NotFound => None,
+                        other_error => {
+                            bail!(
+                                "unable to obtain symlink metadata for file {:?}: {:?}",
+                                PathBuf::from(&clean_path).join(&p).join(&part_path),
+                                other_error
+                            );
+                        }
+                    },
+                };
+
+                if let Some(metadata) = metadata {
+                    if metadata.file_type().is_symlink() {
+                        let link_path =
+                            fs::read_link(PathBuf::from(&clean_path).join(&p).join(&part_path))
```

Comment:
You can use `?` here instead of `unwrap`? We want to return error here instead of `unwrap`.